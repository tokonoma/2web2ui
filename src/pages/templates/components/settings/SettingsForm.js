/* eslint-disable max-lines */
import React from 'react';
import { Field } from 'redux-form';
import { Button, Panel, Stack } from 'src/components/matchbox';
import { ToggleBlock } from 'src/components/toggleBlock';
import SubaccountSection from 'src/components/subaccountSection';
import { TextFieldWrapper } from 'src/components';
import FromEmailWrapper from '../FromEmailWrapper';
import CopyField from 'src/components/copyField/CopyField';
import { required } from 'src/helpers/validation';
import styles from './SettingsForm.module.scss';

export default function SettingsForm(props) {
  const updateSettings = values => {
    const {
      draft,
      updateDraft,
      parsedTestData,
      showAlert,
      content: { amp_html, html, text },
      setHasSaved,
    } = props;

    return updateDraft(
      {
        id: draft.id,
        name: values.name,
        description: values.description,
        // must include other content parts to avoid work in progress
        content: { ...values.content, amp_html, html, text },
        options: values.options,
        // this value is only editable and present when template is not assigned to a subaccount
        shared_with_subaccounts: draft.subaccount_id ? undefined : values.shared_with_subaccounts,
        parsedTestData,
      },
      draft.subaccount_id, // not in form and not allowed to change
    ).then(() => {
      setHasSaved(true);
      showAlert({ type: 'success', message: 'Template settings updated.' });
    });
  };

  const parseToggle = value => !!value;

  const renderPublishedIntro = () => {
    const { hasDraft } = props;
    return (
      <Panel.Section>
        <p
          className={styles.SettingsIntro}
        >{`Template settings can only be changed in drafts. Simply select '${
          hasDraft ? 'Edit Draft' : 'Save as Draft'
        }' in the top right to access the draft version, and adjust settings as needed.`}</p>
      </Panel.Section>
    );
  };

  const {
    handleSubmit,
    domainsLoading,
    domains,
    subaccountId,
    submitting,
    pristine,
    valid,
    hasSubaccounts,
    canViewSubaccount,
    isPublishedMode,
    draft,
    isReadOnly,
  } = props;
  const isEditingDisabled = submitting || isReadOnly;
  const isSubmissionDisabled = submitting || !valid || pristine || isReadOnly;
  const canViewSubaccountSection = hasSubaccounts && canViewSubaccount;
  const fromEmailHelpText =
    !domainsLoading && !domains.length
      ? subaccountId
        ? 'The selected subaccount does not have any verified sending domains.'
        : 'You do not have any verified sending domains to use.'
      : null;

  return (
    <>
      {isPublishedMode && renderPublishedIntro()}
      <form onSubmit={handleSubmit(updateSettings)}>
        <Panel.Section>
          <Stack>
            <Field
              name="name"
              component={TextFieldWrapper}
              label="Template Name"
              disabled={isEditingDisabled}
              validate={required}
            />

            <CopyField
              name="id"
              id="template-id-field"
              label="Template ID"
              value={draft.id}
              helpText={"A Unique ID for your template, we'll fill this in for you."}
              disabled={true}
            />
          </Stack>
        </Panel.Section>

        {canViewSubaccountSection && (
          <Panel.Section>
            <SubaccountSection newTemplate={false} disabled={isEditingDisabled} />
          </Panel.Section>
        )}

        <Panel.Section>
          <Stack>
            <Field
              name="content.subject"
              component={TextFieldWrapper}
              label="Subject"
              validate={required}
              disabled={isEditingDisabled}
            />

            <Field
              name="content.from.email"
              component={FromEmailWrapper}
              placeholder="example@email.com"
              label="From Email"
              // Do not try to validate email, let our API make that decision
              validate={[required]}
              domains={domains}
              helpText={fromEmailHelpText}
              disabled={isEditingDisabled}
            />

            <Field
              name="content.from.name"
              component={TextFieldWrapper}
              label="From Name"
              helpText="A friendly from for your recipients."
              disabled={isEditingDisabled}
            />

            <Field
              // Do not try to validate email, let our API make that decision
              name="content.reply_to"
              component={TextFieldWrapper}
              label="Reply To"
              helpText="An email address recipients can reply to."
              disabled={isEditingDisabled}
            />

            <Field
              name="description"
              component={TextFieldWrapper}
              label="Description"
              helpText="Not visible to recipients."
              disabled={isEditingDisabled}
            />
          </Stack>
        </Panel.Section>
        <Panel.Section>
          <Stack>
            <Field
              name="options.open_tracking"
              component={ToggleBlock}
              label="Track Opens"
              type="checkbox"
              parse={parseToggle}
              disabled={isEditingDisabled}
            />

            <Field
              name="options.click_tracking"
              component={ToggleBlock}
              label="Track Clicks"
              type="checkbox"
              parse={parseToggle}
              disabled={isEditingDisabled}
            />

            <Field
              name="options.transactional"
              component={ToggleBlock}
              label="Transactional"
              type="checkbox"
              parse={parseToggle}
              helpText="Transactional messages are triggered by a userâ€™s actions on the website, like requesting a password reset, signing up, or making a purchase."
              disabled={isEditingDisabled}
            />
          </Stack>
        </Panel.Section>

        <Panel.Section>
          <Button variant="primary" type="submit" disabled={isSubmissionDisabled}>
            Update Settings
          </Button>
        </Panel.Section>
      </form>
    </>
  );
}
