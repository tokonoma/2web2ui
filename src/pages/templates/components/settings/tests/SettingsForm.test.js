import { shallow } from 'enzyme';
import React from 'react';
import cases from 'jest-in-case';
import every from 'lodash/every';
import SettingsForm from '../SettingsForm';

describe('SettingsForm', () => {
  const subject = props => {
    const defaultProps = {
      domains: [{ domain: 'test.com' }, { domain: 'verified.com' }],
      draft: {
        id: 'fake-id',
      },
      domainsLoading: false,
      hasSubaccounts: false,
      canViewSubaccount: false,
      content: {
        html: '<p>Hello</p>',
        text: '',
      },
      values: {
        content: {},
        name: 'Test Template',
      },
      isPublishedMode: false,
      handleSubmit: jest.fn(func => func),
      showAlert: jest.fn(),
      setHasSaved: jest.fn(),
      isReadOnly: false,
    };

    return shallow(<SettingsForm {...defaultProps} {...props} />);
  };

  it('renders without subaccounts', () => {
    const wrapper = subject();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.exists('SubaccountSection')).toBe(false);
  });

  it('renders with subaccounts', () => {
    expect(
      subject({ hasSubaccounts: true, canViewSubaccount: true }).exists('SubaccountSection'),
    ).toBe(true);
  });

  it('renders From Email help text with no verified domains', () => {
    const wrapper = subject({ domains: [] });
    expect(wrapper.find('[name="content.from.email"]').prop('helpText')).toEqual(
      'You do not have any verified sending domains to use.',
    );
  });

  it('renders From Email help text with no verified domains for subaccount', () => {
    const wrapper = subject({ subaccountId: 101, domains: [] });
    expect(wrapper.find('[name="content.from.email"]').prop('helpText')).toEqual(
      'The selected subaccount does not have any verified sending domains.',
    );
  });

  it('renders id field disabled and leverages the CopyField component', () => {
    const wrapper = subject();

    expect(wrapper.find('[name="id"]').prop('disabled')).toBe(true);
    expect(wrapper.find('CopyField')).toEqual(wrapper.find('[name="id"]'));
  });

  describe('Published version', () => {
    it('renders with fields disabled', () => {
      const wrapper = subject({
        hasSubaccounts: true,
        canViewSubaccount: true,
        isReadOnly: true,
      });
      const fieldProps = wrapper.find('Field').map(field => field.prop('disabled'));
      expect(fieldProps.length).toEqual(9);
      expect(every(fieldProps)).toBe(true);
      expect(wrapper.find('SubaccountSection').prop('disabled')).toBe(true);
      expect(wrapper.find('[name="id"]')).toHaveProp('disabled', true);
    });

    it('renders settings intro when draft does not exist', () => {
      expect(
        subject({ hasDraft: false, isPublishedMode: true }).find('[className="SettingsIntro"]'),
      ).toMatchSnapshot();
    });

    it('renders settings intro when draft exists', () => {
      expect(
        subject({ hasDraft: true, isPublishedMode: true }).find('[className="SettingsIntro"]'),
      ).toMatchSnapshot();
    });
  });

  cases(
    'disables submit button',
    ({ name, ...props }) => {
      const wrapper = subject({ valid: true, ...props });
      expect(wrapper.find('Button[type="submit"]')).toHaveProp('disabled', true);
    },
    {
      'when submitting': {
        submitting: true,
      },
      'when pristine': {
        pristine: true,
      },
      'when invalid': {
        valid: false,
      },
      'when in publish mode': {
        isReadOnly: true,
      },
    },
  );

  describe('when form is submitted', () => {
    const submitForm = async (props = {}, values = {}) => {
      const wrapper = subject({
        parsedTestData: {
          name: 'Bob Bell',
        },
        pristine: false,
        setHasSaved: () => {},
        showAlert: () => {},
        updateDraft: () => Promise.resolve(),
        valid: true,
        ...props,
      });

      await wrapper.find('form').simulate('submit', {
        description: '',
        options: { click_tracking: true, open_tracking: true, transactional: true },
        ...values,
        content: { text: '', ...values.content },
      });

      return wrapper;
    };

    it('requests setting update', async () => {
      const draft = { id: 'test-template' };
      const content = { html: 'Hello {{name}}', text: 'Hello {{name}}' };
      const values = {
        name: 'Test Template',
        description: 'This is only a test',
        shared_with_subaccounts: true,
        content: {
          from: { email: 'bob@bell.com' },
        },
      };
      const updateDraft = jest.fn(() => Promise.resolve());
      await submitForm({ content, draft, updateDraft }, values);

      expect(updateDraft).toHaveBeenCalledWith(
        {
          id: 'test-template',
          name: 'Test Template',
          description: 'This is only a test',
          content: {
            amp_html: undefined,
            html: 'Hello {{name}}',
            text: 'Hello {{name}}',
            from: { email: 'bob@bell.com' },
          },
          options: { click_tracking: true, open_tracking: true, transactional: true },
          shared_with_subaccounts: true,
          parsedTestData: {
            name: 'Bob Bell',
          },
        },
        undefined,
      );
    });

    it('requests setting update for template assinged to subaccount', async () => {
      const draft = {
        id: 'test-template-for-123',
        subaccount_id: 123,
      };
      const updateDraft = jest.fn(() => Promise.resolve());
      await submitForm({ draft, updateDraft });

      expect(updateDraft).toHaveBeenCalledWith(
        expect.objectContaining({
          id: 'test-template-for-123',
          shared_with_subaccounts: undefined,
        }),
        123,
      );
    });

    it('reports have has completed', async () => {
      const setHasSaved = jest.fn();
      await submitForm({ setHasSaved });

      expect(setHasSaved).toHaveBeenCalledWith(true);
    });

    it('alerts on success', async () => {
      const showAlert = jest.fn();
      await submitForm({ showAlert });

      expect(showAlert).toHaveBeenCalledWith({
        type: 'success',
        message: 'Template settings updated.',
      });
    });
  });
});
