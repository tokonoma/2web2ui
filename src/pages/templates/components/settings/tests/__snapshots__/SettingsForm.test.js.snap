// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsForm Published version renders settings intro when draft does not exist 1`] = `
<p
  className="SettingsIntro"
>
  Template settings can only be changed in drafts. Simply select 'Save as Draft' in the top right to access the draft version, and adjust settings as needed.
</p>
`;

exports[`SettingsForm Published version renders settings intro when draft exists 1`] = `
<p
  className="SettingsIntro"
>
  Template settings can only be changed in drafts. Simply select 'Edit Draft' in the top right to access the draft version, and adjust settings as needed.
</p>
`;
